### PR TITLE
Optimization for eliminating redundant memory operations

### PR DESCRIPTION
**Pull request by yyzdtccjdtc**
Optimization for eliminating redundant memory operations related to this issue.
https://github.com/sharkdp/pastel/issues/163#issue-1223375579

### CODE REVIEW

**Comment by yyzdtccjdtc** [comment]
> Nice. Can you tell us a bit more about how you benchmarked this?

I tested it with the command line 'pastel distinct 2000 >/dev/null' and then took the average execution time. According to my test results, the running time was reduced from 9.54 seconds to 8.41 seconds, which has a 13.4% speedup.

Also I compiled it with the opt-level = 3.

**Comment by yyzdtccjdtc** [comment]
> Unfortunately, I can not reproduce these benchmark results:
> 
> Command	Mean [s]	Min [s]	Max [s]	Relative
> `./pastel-master distinct 200`	1.166 ± 0.023	1.141	1.208	1.00
> `./pastel-165 distinct 200`	1.252 ± 0.023	1.215	1.298	1.07 ± 0.03
> These results were created using a normal `cargo build --release` build. Benchmarking was done with hyperfine:
> 
> ```
> hyperfine -L version master,165 --warmup 3 './pastel-{version} distinct 200' --export-markdown results.md
> ```

I believe if you try to change it to distinct 2000, you can see the difference.

**Comment by yyzdtccjdtc** [comment]
By carefully looking into the code I found the problem. The Lab structures in the `Vec<(Color, Lab)>` are generated here. 

https://github.com/sharkdp/pastel/blob/47f9ddd59ea14df8d9d1f1073d14501751a49fd0/src/distinct.rs#L82-L85

So my get_labs function to get the Lab structure out is a redundant operation.

```
    pub fn get_labs(&self) -> Vec<Lab> {
        self.colors.iter().map(|(_, l)| l.clone()).collect()
    }
```

My current approach is to generate two separate labs and colors vectors directly at the time of this SimulatedAnnealing structure creation.

According to my tests with '/pastel distinct 200 >results.md', the average execution time decreased from 0.7158s to 0.6389s, which is a 12% speedup. Hopefully, your test will achieve the same results as mine :)

File: src/distinct.rs [hunk 1]
```diff
@@ -65,7 +65,8 @@ pub struct SimulationParameters {
 }
 
 pub struct SimulatedAnnealing<R: Rng> {
-    colors: Vec<(Color, Lab)>,
+    colors: Vec<Color>,
+    labs: Vec<Lab>,
```
Comment (sharkdp) [inline src/distinct.rs:69]: Can we maybe call it `lab_values` everywhere, instead of `labs`?

File: src/lib.rs [hunk 2]
```diff
@@ -938,7 +938,7 @@ impl fmt::Display for LMS {
     }
 }
 
-#[derive(Debug, Clone, PartialEq)]
+#[derive(Debug, Clone, PartialEq, Copy)]
```
Comment (sharkdp) [inline src/lib.rs:941]: Do we really want to derive copy? I'd prefer if we explicitly call `.clone()` when we really want it. The struct contains 4 64bit floats, so it's *probably* cheaper to usually pass it by reference, not by value, right?

File: src/distinct.rs [hunk 3]
```diff
@@ -377,10 +377,10 @@ impl DistanceResult {
             (self.closest_distances.len() - self.num_fixed_colors) as Scalar;
     }
 
-    fn distance(&self, a: &(Color, Lab), b: &(Color, Lab)) -> Scalar {
+    fn distance(&self, a: &Lab, b: &Lab) -> Scalar {
         match self.distance_metric {
-            DistanceMetric::CIE76 => delta_e::cie76(&a.1, &b.1),
-            DistanceMetric::CIEDE2000 => delta_e::ciede2000(&a.1, &b.1),
+            DistanceMetric::CIE76 => delta_e::cie76(&a, &b),
```
Comment (sharkdp) [inline src/distinct.rs:382]: Could you please fix the clippy warning here: "this expression creates a reference which is immediately dereferenced by the compiler"

File: src/distinct.rs [hunk 4]
```diff
@@ -377,10 +377,10 @@ impl DistanceResult {
...
+            DistanceMetric::CIEDE2000 => delta_e::ciede2000(&a, &b),
```
Comment (sharkdp) [inline src/distinct.rs:383]: … and here.

**Comment by sharkdp** [comment]
Now I can reproduce those results - thank you!

**Comment by yyzdtccjdtc** [comment]
I changed all labs to lab_values and fixed the clippy warning. I also replaced the derive copy with clone() function at the same time.

`The struct contains 4 64bit floats, so it's probably cheaper to usually pass it by reference, not by value, right?`

I also looked into this issue a bit. According to the output file of objdump, I found that if we use the reference like this 
`let at_lab = &lab_values[color]`, it will load the address of 'lab_values[color]' first, then push this address into the stack, and finally it will first load the address from the stack and then load the value from this address to do the calculation.

But if we use copy() or clone() `let at_lab = lab_values[color].clone();`, it will load the values directly and push it into the stack, and then take the values  from the stack to do the calculation where it is needed. So clone() has one less load operation compared to reference.

And the execution time also proves this, the average time of reference is 0.68s (1.06x speedup), while the average time of clone is 0.64s (1.12x speedup).

**Comment by yyzdtccjdtc** [comment]
@sharkdp Please tell me if there are any other changes I need to make?

**Comment by sharkdp** [comment]
No, looks good. Thank you very much!

**Pull request merged**